File: app/src/main/java/me/zhanghai/android/files/provider/root/RootableFileSystemProvider.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.root;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.URI;
import java.util.List;
import java.util.Map;
import java.util.Set;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java8.nio.channels.FileChannel;
import java8.nio.channels.SeekableByteChannel;
import java8.nio.file.AccessMode;
import java8.nio.file.CopyOption;
import java8.nio.file.DirectoryStream;
import java8.nio.file.FileStore;
import java8.nio.file.FileSystem;
import java8.nio.file.LinkOption;
import java8.nio.file.OpenOption;
import java8.nio.file.Path;
import java8.nio.file.attribute.BasicFileAttributes;
import java8.nio.file.attribute.FileAttribute;
import java8.nio.file.attribute.FileAttributeView;
import java8.nio.file.spi.FileSystemProvider;
import java9.util.function.Consumer;
import java9.util.function.Function;
import me.zhanghai.android.files.provider.common.PathObservable;
import me.zhanghai.android.files.provider.common.PathObservableProvider;
import me.zhanghai.android.files.provider.common.Searchable;

public class RootableFileSystemProvider extends FileSystemProvider
        implements PathObservableProvider, Searchable {

    @NonNull
    private final FileSystemProvider mLocalProvider;
    @NonNull
    private final FileSystemProvider mRootProvider;

    public RootableFileSystemProvider(
            @NonNull Function<FileSystemProvider, FileSystemProvider> newLocalProvider,
            @NonNull Function<FileSystemProvider, FileSystemProvider> newRootProvider) {
        mLocalProvider = newLocalProvider.apply(this);
        mRootProvider = newRootProvider.apply(this);
    }

    @NonNull
    protected <FSP extends FileSystemProvider> FSP getLocalProvider() {
        //noinspection unchecked
        return (FSP) mLocalProvider;
    }

    @NonNull
    protected <FSP extends FileSystemProvider> FSP getRootProvider() {
        //noinspection unchecked
        return (FSP) mRootProvider;
    }

    @NonNull
    @Override
    public String getScheme() {
        return mLocalProvider.getScheme();
    }

    @NonNull
    @Override
    public FileSystem newFileSystem(@NonNull URI uri, @NonNull Map<String, ?> env)
            throws IOException {
        return mLocalProvider.newFileSystem(uri, env);
    }

    @NonNull
    @Override
    public FileSystem getFileSystem(@NonNull URI uri) {
        return mLocalProvider.getFileSystem(uri);
    }

    @NonNull
    @Override
    public Path getPath(@NonNull URI uri) {
        return mLocalProvider.getPath(uri);
    }

    @Override
    public InputStream newInputStream(Path path, OpenOption... options) throws IOException {
        return applyRootable(path, provider -> provider.newInputStream(path, options));
    }

    @Override
    public OutputStream newOutputStream(Path path, OpenOption... options) throws IOException {
        return applyRootable(path, provider -> provider.newOutputStream(path, options));
    }

    @NonNull
    @Override
    public FileChannel newFileChannel(@NonNull Path path,
                                      @NonNull Set<? extends OpenOption> options,
                                      @NonNull FileAttribute<?>... attributes) throws IOException {
        return applyRootable(path, provider -> provider.newFileChannel(path, options, attributes));
    }

    @NonNull
    @Override
    public SeekableByteChannel newByteChannel(@NonNull Path path,
                                              @NonNull Set<? extends OpenOption> options,
                                              @NonNull FileAttribute<?>... attributes)
            throws IOException {
        return applyRootable(path, provider -> provider.newByteChannel(path, options, attributes));
    }

    @NonNull
    @Override
    public DirectoryStream<Path> newDirectoryStream(
            @NonNull Path directory, @NonNull DirectoryStream.Filter<? super Path> filter)
            throws IOException {
        return applyRootable(directory, provider -> provider.newDirectoryStream(directory, filter));
    }

    @Override
    public void createDirectory(@NonNull Path directory, @NonNull FileAttribute<?>... attributes)
            throws IOException {
        acceptRootable(directory, provider -> provider.createDirectory(directory, attributes));
    }

    @Override
    public void createSymbolicLink(@NonNull Path link, @NonNull Path target,
                                   @NonNull FileAttribute<?>... attributes) throws IOException {
        acceptRootable(link, target, provider -> provider.createSymbolicLink(link, target,
                attributes));
    }

    @Override
    public void createLink(Path link, Path existing) throws IOException {
        acceptRootable(link, existing, provider -> provider.createLink(link, existing));
    }

    @Override
    public void delete(@NonNull Path path) throws IOException {
        acceptRootable(path, provider -> provider.delete(path));
    }

    @Override
    public Path readSymbolicLink(Path link) throws IOException {
        return applyRootable(link, provider -> provider.readSymbolicLink(link));
    }

    @Override
    public void copy(@NonNull Path source, @NonNull Path target, @NonNull CopyOption... options)
            throws IOException {
        acceptRootable(source, target, provider -> provider.copy(source, target, options));
    }

    @Override
    public void move(@NonNull Path source, @NonNull Path target, @NonNull CopyOption... options)
            throws IOException {
        acceptRootable(source, target, provider -> provider.move(source, target, options));
    }

    @Override
    public boolean isSameFile(@NonNull Path path, @NonNull Path path2) throws IOException {
        return applyRootable(path, path2, provider -> provider.isSameFile(path, path2));
    }

    @Override
    public boolean isHidden(@NonNull Path path) throws IOException {
        return applyRootable(path, provider -> provider.isHidden(path));
    }

    @NonNull
    @Override
    public FileStore getFileStore(@NonNull Path path) throws IOException {
        return applyRootable(path, provider -> provider.getFileStore(path));
    }

    @Override
    public void checkAccess(@NonNull Path path, @NonNull AccessMode... modes) throws IOException {
        acceptRootable(path, provider -> provider.checkAccess(path, modes));
    }

    @Nullable
    @Override
    public <V extends FileAttributeView> V getFileAttributeView(@NonNull Path path,
                                                                @NonNull Class<V> type,
                                                                @NonNull LinkOption... options) {
        return mLocalProvider.getFileAttributeView(path, type, options);
    }

    @NonNull
    @Override
    public <A extends BasicFileAttributes> A readAttributes(@NonNull Path path,
                                                            @NonNull Class<A> type,
                                                            @NonNull LinkOption... options)
            throws IOException {
        return mLocalProvider.readAttributes(path, type, options);
    }

    @NonNull
    @Override
    public Map<String, Object> readAttributes(@NonNull Path path, @NonNull String attributes,
                                              @NonNull LinkOption... options) throws IOException {
        return applyRootable(path, provider -> provider.readAttributes(path, attributes, options));
    }

    @Override
    public void setAttribute(@NonNull Path path, @NonNull String attribute, Object value,
                             @NonNull LinkOption... options) throws IOException {
        acceptRootable(path, provider -> provider.setAttribute(path, attribute, value, options));
    }

    @NonNull
    @Override
    public PathObservable observePath(@NonNull Path path, long intervalMillis) throws IOException {
        if (!(mLocalProvider instanceof PathObservableProvider)) {
            throw new UnsupportedOperationException();
        }
        return applyRootable(path, provider -> {
            // observePath() may or may not be able to detect denied access, and that is expansive
            // on Linux (having to create the WatchService first before registering a WatchKey). So
            // we check the access beforehand.
            if (provider == mLocalProvider) {
                BasicFileAttributes attributes = null;
                try {
                    attributes = provider.readAttributes(path, BasicFileAttributes.class);
                } catch (IOException ignored) {}
                if (attributes == null) {
                    attributes = provider.readAttributes(path, BasicFileAttributes.class,
                            LinkOption.NOFOLLOW_LINKS);
                }
                if (attributes.isSymbolicLink()) {
                    provider.readSymbolicLink(path);
                } else {
                    provider.checkAccess(path, AccessMode.READ);
                }
            }
            return ((PathObservableProvider) provider).observePath(path, intervalMillis);
        });
    }

    @Override
    public void search(@NonNull Path directory, @NonNull String query,
                       @NonNull Consumer<List<Path>> listener, long intervalMillis)
            throws IOException {
        acceptRootable(directory, provider -> ((Searchable) provider).search(directory, query,
                listener, intervalMillis));
    }

    private void acceptRootable(@NonNull Path path,
                                RootUtils.Consumer<FileSystemProvider> consumer)
            throws IOException {
        RootUtils.acceptRootable(path, mLocalProvider, mRootProvider, consumer);
    }

    private void acceptRootable(@NonNull Path path1, @NonNull Path path2,
                                RootUtils.Consumer<FileSystemProvider> consumer)
            throws IOException {
        RootUtils.acceptRootable(path1, path2, mLocalProvider, mRootProvider, consumer);
    }

    private <R> R applyRootable(@NonNull Path path,
                                 RootUtils.Function<FileSystemProvider, R> function)
            throws IOException {
        return RootUtils.applyRootable(path, mLocalProvider, mRootProvider, function);
    }

    private <R> R applyRootable(@NonNull Path path1, @NonNull Path path2,
                                @NonNull RootUtils.Function<FileSystemProvider, R> function)
            throws IOException {
        return RootUtils.applyRootable(path1, path2, mLocalProvider, mRootProvider, function);
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/root/RootableFileSystemProvider.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.root

import java8.nio.channels.FileChannel
import java8.nio.channels.SeekableByteChannel
import java8.nio.file.AccessMode
import java8.nio.file.CopyOption
import java8.nio.file.DirectoryStream
import java8.nio.file.FileStore
import java8.nio.file.FileSystem
import java8.nio.file.LinkOption
import java8.nio.file.OpenOption
import java8.nio.file.Path
import java8.nio.file.attribute.BasicFileAttributes
import java8.nio.file.attribute.FileAttribute
import java8.nio.file.attribute.FileAttributeView
import java8.nio.file.spi.FileSystemProvider
import me.zhanghai.android.files.provider.common.PathObservable
import me.zhanghai.android.files.provider.common.PathObservableProvider
import me.zhanghai.android.files.provider.common.Searchable
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.URI

abstract class RootableFileSystemProvider(
    localProviderCreator: (FileSystemProvider) -> FileSystemProvider,
    rootProviderCreator: (FileSystemProvider) -> FileSystemProvider
) : FileSystemProvider(), PathObservableProvider, Searchable {
    protected open val localProvider: FileSystemProvider = localProviderCreator(this)
    protected open val rootProvider: FileSystemProvider = rootProviderCreator(this)

    override fun getScheme(): String = localProvider.scheme

    @Throws(IOException::class)
    override fun newFileSystem(uri: URI, env: Map<String, *>): FileSystem =
        localProvider.newFileSystem(uri, env)

    override fun getFileSystem(uri: URI): FileSystem = localProvider.getFileSystem(uri)

    override fun getPath(uri: URI): Path = localProvider.getPath(uri)

    @Throws(IOException::class)
    override fun newInputStream(path: Path, vararg options: OpenOption): InputStream =
        callRootable(path) { newInputStream(path, *options) }

    @Throws(IOException::class)
    override fun newOutputStream(path: Path, vararg options: OpenOption): OutputStream =
        callRootable(path) { newOutputStream(path, *options) }

    @Throws(IOException::class)
    override fun newFileChannel(
        path: Path,
        options: Set<OpenOption>,
        vararg attributes: FileAttribute<*>
    ): FileChannel = callRootable(path) { newFileChannel(path, options, *attributes) }

    @Throws(IOException::class)
    override fun newByteChannel(
        path: Path,
        options: Set<OpenOption>,
        vararg attributes: FileAttribute<*>
    ): SeekableByteChannel = callRootable(path) { newByteChannel(path, options, *attributes) }

    @Throws(IOException::class)
    override fun newDirectoryStream(
        directory: Path,
        filter: DirectoryStream.Filter<in Path>
    ): DirectoryStream<Path> = callRootable(directory) { newDirectoryStream(directory, filter) }

    @Throws(IOException::class)
    override fun createDirectory(directory: Path, vararg attributes: FileAttribute<*>) {
        callRootable(directory) { createDirectory(directory, *attributes) }
    }

    @Throws(IOException::class)
    override fun createSymbolicLink(link: Path, target: Path, vararg attributes: FileAttribute<*>) {
        callRootable(link, target) { createSymbolicLink(link, target, *attributes) }
    }

    @Throws(IOException::class)
    override fun createLink(link: Path, existing: Path) {
        callRootable(link, existing) {createLink(link, existing) }
    }

    @Throws(IOException::class)
    override fun delete(path: Path) {
        callRootable(path) { delete(path) }
    }

    @Throws(IOException::class)
    override fun readSymbolicLink(link: Path): Path = callRootable(link) { readSymbolicLink(link) }

    @Throws(IOException::class)
    override fun copy(source: Path, target: Path, vararg options: CopyOption) {
        callRootable(source, target) { copy(source, target, *options) }
    }

    @Throws(IOException::class)
    override fun move(source: Path, target: Path, vararg options: CopyOption) {
        callRootable(source, target) { move(source, target, *options) }
    }

    @Throws(IOException::class)
    override fun isSameFile(path: Path, path2: Path): Boolean =
        callRootable(path, path2) { isSameFile(path, path2) }

    @Throws(IOException::class)
    override fun isHidden(path: Path): Boolean = callRootable(path) { isHidden(path) }

    @Throws(IOException::class)
    override fun getFileStore(path: Path): FileStore = callRootable(path) { getFileStore(path) }

    @Throws(IOException::class)
    override fun checkAccess(path: Path, vararg modes: AccessMode) {
        callRootable(path) {checkAccess(path, *modes) }
    }

    override fun <V : FileAttributeView> getFileAttributeView(
        path: Path,
        type: Class<V>,
        vararg options: LinkOption
    ): V? = localProvider.getFileAttributeView(path, type, *options)

    @Throws(IOException::class)
    override fun <A : BasicFileAttributes> readAttributes(
        path: Path,
        type: Class<A>,
        vararg options: LinkOption
    ): A = localProvider.readAttributes(path, type, *options)

    @Throws(IOException::class)
    override fun readAttributes(
        path: Path,
        attributes: String,
        vararg options: LinkOption
    ): Map<String, Any> = callRootable(path) { readAttributes(path, attributes, *options) }

    @Throws(IOException::class)
    override fun setAttribute(
        path: Path,
        attribute: String,
        value: Any,
        vararg options: LinkOption
    ) {
        callRootable(path) { setAttribute(path, attribute, value, *options) }
    }

    @Throws(IOException::class)
    override fun observe(path: Path, intervalMillis: Long): PathObservable {
        if (localProvider !is PathObservableProvider) {
            throw UnsupportedOperationException()
        }
        return callRootable(path) {
            // observe() may or may not be able to detect denied access, and that is expensive on
            // Linux (having to create the WatchService first before registering a WatchKey). So we
            // check the access beforehand.
            if (this == localProvider) {
                val attributes = try {
                    readAttributes(path, BasicFileAttributes::class.java)
                } catch (ignored: IOException) {
                    readAttributes(path, BasicFileAttributes::class.java, LinkOption.NOFOLLOW_LINKS)
                }
                if (attributes.isSymbolicLink) {
                    readSymbolicLink(path)
                } else {
                    checkAccess(path, AccessMode.READ)
                }
            }
            (this as PathObservableProvider).observe(path, intervalMillis)
        }
    }

    @Throws(IOException::class)
    override fun search(
        directory: Path,
        query: String,
        intervalMillis: Long,
        listener: (List<Path>) -> Unit
    ) {
        callRootable(directory) {
            (this as Searchable).search(directory, query, intervalMillis, listener)
        }
    }

    @Throws(IOException::class)
    private fun <R> callRootable(path: Path, block: FileSystemProvider.() -> R): R =
        callRootable(path, localProvider, rootProvider, block)

    @Throws(IOException::class)
    private fun <R> callRootable(path1: Path, path2: Path, block: FileSystemProvider.() -> R): R =
        callRootable(path1, path2, localProvider, rootProvider, block)
}
